Search for GitLab PRs in Launchpad

sergeibbb opened the pull request:
#3788
# Description


# Checklist

<!-- Please check off the following -->

- [x] I have followed the guidelines in the Contributing document
- [x] My changes follow the coding style of this project
- [x] My changes build without any errors or warnings
- [x] My changes have been formatted and linted
- [x] My changes include any required corresponding changes to the documentation (including CHANGELOG.md and README.md)
- [x] My changes have been rebased and squashed to the minimal number (typically 1) of relevant commits
- [x] My changes have a descriptive commit message with a short title, including a `Fixes $XXX -` or `Closes #XXX -` prefix to auto-close the issue that your PR addresses

sergeibbb commented:
@axosoft-ramint 

In the last commit I'm grouping detected PR-identity by providers, because thought that if we know that an URL matches to a GitLab then no need to search for it in GitHub. But if it looks too compicated, then then last commit can be rejected.

axosoft-ramint reviewed (changes requested):
As it stands, this doesn't quite feel ready. The utils and test files that are provider-specific should be organized into provider-specific directories. The url identity being a mapped object by provider id feels confusing - the identity itself should just come in on the function. Some of the changes to the PR utils need either explanation or rewrite to make more sense (why are we defining an undefined const and then returning it without touching it?). The same function being defined in the PR utils but then also implemented specifically for integrations doesn't quite make sense to me. The `checkoutable` logic in launchpad shouldn't be necessary - we should always have at least a defined head ref and remote url for a PR, or else we're not fetching enough info.

Can talk on this further on your return.

sergeibbb commented:
@axosoft-ramint 

> The utils and test files that are provider-specific should be organized into provider-specific directories.

fixed

> The url identity being a mapped object by provider id feels confusing - the identity itself should just come in on the function.

fixed

> why are we defining an undefined const and then returning it without touching it?

where?

> The same function being defined in the PR utils but then also implemented specifically for integrations doesn't quite make sense to me

I access the necessary implementation of the function through the integration instance that's why it's in method.
But I cannot keep the implementation in the integration because with our current test configuration it cannot be tested (our tests cannot handle importing the Context). To make the function testable as unit I move it to an external file.

> The checkoutable logic in launchpad shouldn't be necessary

I'll try to fix it. I think that in order to do that I have to rewrite retrieving GitLab PRs from REST to GraphQL.

sergeibbb commented:
@axosoft-ramint 

> The utils and test files that are provider-specific should be organized into provider-specific directories.

fixed

> The url identity being a mapped object by provider id feels confusing - the identity itself should just come in on the function.

fixed

> why are we defining an undefined const and then returning it without touching it?

where?

> The same function being defined in the PR utils but then also implemented specifically for integrations doesn't quite make sense to me

I access the necessary implementation of the function through the integration instance that's why it's in method.
But I cannot keep the implementation in the integration because with our current test configuration it cannot be tested (our tests cannot handle importing the Context). To make the function testable as unit I move it to an external file.

> The checkoutable logic in launchpad shouldn't be necessary

I'll try to fix it. I think that in order to do that I have to rewrite retrieving GitLab PRs from REST to GraphQL.

But initially I thought that it can be a follow-up:

<img width="800" alt="image" src="https://github.com/user-attachments/assets/4137d897-cc70-4a5d-a109-9751e0294dc2" />

sergeibbb commented:
@axosoft-ramint 

> The utils and test files that are provider-specific should be organized into provider-specific directories.

fixed

> The url identity being a mapped object by provider id feels confusing - the identity itself should just come in on the function.

fixed

> why are we defining an undefined const and then returning it without touching it?

fixed

> The same function being defined in the PR utils but then also implemented specifically for integrations doesn't quite make sense to me

I access the necessary implementation of the function through the integration instance that's why it's in method.
But I cannot keep the implementation in the integration because with our current test configuration it cannot be tested (our tests cannot handle importing the Context). To make the function testable as unit I move it to an external file.

But I'm not sure about one thing: https://github.com/gitkraken/vscode-gitlens/blob/89bd972aa29c196a2b179e7a08ffad45291913ab/src/plus/launchpad/launchpadProvider.ts#L584-L611


> The checkoutable logic in launchpad shouldn't be necessary

Fixed.

axosoft-ramint commented on src/git/models/pullRequest.ts:
```diff
@@ -272,7 +270,7 @@ export function getRepositoryIdentityForPullRequest(
 		return {
 			remote: {
-				url: pr.refs?.head?.url,
+				url: pr.repository.url,
```
This changes the url ref to the base repository. This would break the model for PRs where the PR is targeting a different repository and (like when the head ref is from someone's fork).

Commands that use this like "open the PR's repository" or "open in worktree" will break in those cases.

sergeibbb replied:
@axosoft-ramint 
Thank you for catching it.

Question here. `pr.refs?.head?.url` is a link to a branch to to the repo. It fails matching to `uniqueRemoteUrls` in the LaunchpadProvider when I try to `swtichTo` the branch.
So I see 3 ways to fix it:

1. Add `pr.refs.head.repoUrl` on the specific provider level.
2. Fix the matching to `uniqueRemoteUrls` in launchpadProvider
3. Cut it right here: `url: someOperationOnRefUrl(pr.refs?.head?.url)`

I think that the first option is the best because it can respect specifics of each type of integration. What do you think?

axosoft-ramint replied:
@sergeibbb `pr.hefs?.head?.url` should be the link to the remote (repository) that the head ref is hosted on, not a link to a branch - at least, with GitHub this is the case. Are you finding that with GitLab, it is linking to a branch? If so, whatever we are using to convert GitLab PRs into our PR model, I think we are plugging the wrong datapoint into that property and should start there.

sergeibbb replied:
@axosoft-ramint 

GitLab REST API does not provide necessary information about source and target repositories. So, I've fixed it by switching to GraphQL.  (26727ffce4a3546a1b618d38c7c15e4d615d2a4f) 

However I haven't found a good way to search MRs via GraphQL. So I've solved it with 2 requests. I still use REST to search and then I use PRs' IDs to retrieve the rest of the fields from GraphQL. (3cb31d915a128b051a77e026e7d43791f5fbe0a4) 

cc @nzaytsev

axosoft-ramint commented on src/plus/launchpad/launchpadProvider.ts:
```diff
@@ -210,44 +214,74 @@ export class LaunchpadProvider implements Disposable {
-			if (pr?.value != null) {
-				result = { value: [{ pullRequest: pr.value, reasons: [] }], duration: pr.duration };
-				return { prs: result, suggestionCounts: undefined };
+		const prUrlIdentity: PullRequestUrlIdentity | undefined = await this.getPullRequestIdentityFromMaybeUrl(search);
+		const result: { readonly value: SearchedPullRequest[]; duration: number } = {
+			value: [],
+			duration: 0,
+		};
+
+		const connectedIntegrations = await this.getConnectedIntegrations();
```
If we're already getting the connected integrations below, maybe we can just pass them into `this.getPullRequestIdentityFromMaybeUrl`

sergeibbb replied:
fixed

axosoft-ramint commented on src/plus/launchpad/launchpadProvider.ts:
```diff
@@ -210,44 +214,74 @@ export class LaunchpadProvider implements Disposable {
+		): Promise<undefined | TimedResult<SearchedPullRequest[] | undefined>> => {
 			const prs = await withDurationAndSlowEventOnTimeout(
-				integration?.searchPullRequests(search, undefined, cancellation),
+				integration?.searchPullRequests(search, undefined, cancellation), //
```
Empty comment line at the end

sergeibbb replied:
fixed

axosoft-ramint reviewed (changes requested):
Seems to work so far on my end, but I noticed an issue with the url reference in the RP repo identity that will likely break functionality in some cases with other features.

axosoft-ramint commented on src/plus/integrations/integration.ts:
```diff
@@ -1361,4 +1362,6 @@ export abstract class HostingIntegration<
 		cancellation?: CancellationToken,
 	): Promise<PullRequest[] | undefined>;
+
+	getPullRequestIdentityFromMaybeUrl?(search: string): PullRequestUrlIdentity | undefined;
```
Let's follow a similar pattern for generic and provider-specific function implementation and naming here as we do for other functions, in case we want to expand on the base logic later:

1. Have a `protected getProviderPullRequestIdentityFromMaybeUrl` that can be implemented by specific providers (i.e. GitHub, GitLab, etc.)
2. Have a `getPullRequestIdentityFromMaybeUrl` that, for now, just does `return getProviderPullRequestIdentityFromMaybeUrl` and later if we want, we can layer in caching there.

sergeibbb replied:
fixed

axosoft-ramint commented on src/plus/launchpad/launchpadProvider.ts:
```diff
@@ -547,6 +583,30 @@ export class LaunchpadProvider implements Disposable {
+		);
+	}
+
+	async getPullRequestIdentityFromMaybeUrl(
```
I suggest calling this one `getPullRequestIdentityFromSearch` to avoid making it seem as if line 607 is a recursive call to the same function.

sergeibbb replied:
fixed

axosoft-ramint reviewed (commented):
On testing, the only issue I'm seeing is that owner/repo are not showing up on GitLab search results:

<img width="299" alt="image" src="https://github.com/user-attachments/assets/1c9cb65c-2a8e-4837-846c-da65b7a870eb" />

They do for GitLab PRs that are not search results.

Apart from that, some minor nitpicks that should be easy to change:

sergeibbb commented on src/plus/integrations/integration.ts:
```diff
@@ -556,7 +557,7 @@ export abstract class IntegrationBase<
 		if (!connected) return undefined;
 
-		const pr = this.container.cache.getPullRequest(id, resource, this, () => ({
+		const pr = await this.container.cache.getPullRequest(id, resource, this, () => ({
```
@axosoft-ramint 

I did it because I set a breakpoint at this place it was convenient to see a resolved promise.

sergeibbb commented:
@axosoft-ramint 

> I'm seeing is that owner/repo are not showing up on GitLab search results:

fixed

axosoft-ramint reviewed (approved):
I wont have the chance to test this morning, but if all my prior comments are addressed, feel free to merge.
